### PR TITLE
Tech ensemble model improvements

### PR DESCRIPTION
**Pull request by lizgzil**
**Review notes:**
The main part of this PR (and the bit that is most new) is the analysis in `notebooks/Ensemble parameter exploration.ipynb`.

To just download relevant data and models:
```
make sync_latest_files_from_s3
```
There aren't many tests, but it'd be good to try out `make test`.

**Description**

This PR aims refactors the ensemble model process removing the need for the `ensemble_model.py` script (although kept in).

The algorithm steps are:
1. Get test set predictions for some models (models A, B, C)
2. Create all the combinations of these models ((A, B), (A, C) and (A, B, C))
3. See how well each combination performs as an ensemble also varying the number of models that needs to agree (e.g. using the model combination (A, B, C) if 2 out of 3 models agree a grant is tech, then tag as tech)
4. Decide when combination gives the best performance
5. Use this 'best' ensemble information (e.g. combination= (A, B, C), num_agree = 2) to run future predictions on new grants data

Thus in this PR, I:
- save out test set predictions from models from a particular datestamp `predict_tech.py` (step 1)
- notebook to find best ensemble model `Ensemble parameter exploration.ipynb` (step 2, 3, 4)
- new config file with best ensemble parameters
- ensemble model script `tech_grant_tagger.py` which takes the config file as an argument (step 4)
- tests for the ensemble model script

https://trello.com/c/Pvt0kHiX/1181-rerunning-tech-model-with-new-prodigy-data

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -8,4 +8,5 @@ lxml
 wellcomeml[deep-learning]==2020.11.1
 docutils==0.15.2
 bokeh==2.1.0
-scholarly
\ No newline at end of file
+scholarly
```
Comment (nsorros) [inline requirements.txt:11]: that is an interesting package https://github.com/scholarly-python-package/scholarly, check @aCampello

File: nutrition_labels/prodigy_training_data.py [hunk 2]
```diff
@@ -43,7 +43,8 @@
                 training_data.append(annotation)
 
     training_data = pd.DataFrame(training_data)
-    training_data.drop_duplicates(subset=['Internal ID']).reset_index()
+    training_data.drop_duplicates(subset=['Internal ID'], inplace=True)
```
Comment (nsorros) [inline nutrition_labels/prodigy_training_data.py:46]: this changes the result from before, was there a bug?

File: configs/ensemble/2021.02.20.ini [hunk 3]
```diff
@@ -0,0 +1,12 @@
+[DEFAULT]
```
Comment (nsorros) [inline configs/ensemble/2021.02.20.ini:1]: we love configs. is this the only config?

File: Makefile [hunk 4]
```diff
@@ -63,3 +63,25 @@ sync_models_to_s3:
 .PHONY:sync_models_from_s3
 sync_models_from_s3:
 	aws s3 sync s3://$(PROJECT_BUCKET)/models/ models/
+
+.PHONY:sync_latest_files_to_s3
```
Comment (nsorros) [inline Makefile:67]: what is going on here? why do we go line by line?

File: Makefile [hunk 5]
```diff
@@ -63,3 +63,25 @@ sync_models_to_s3:
...
+sync_latest_files_to_s3:
+	@while read line; do \
+		if [ -d  $$line ]; \
+		then aws s3 cp --recursive $$line s3://$(PROJECT_BUCKET)/$$line; \
+		else aws s3 cp $$line s3://$(PROJECT_BUCKET)/$$line; \
+		fi; \
+	done < latest_files.txt ;
+
+.PHONY:sync_latest_files_from_s3
```
Comment (nsorros) [inline Makefile:76]: is that different from the above? they look identical

File: notebooks/Ensemble parameter exploration.ipynb [hunk 6]
```diff
@@ -0,0 +1,1124 @@
...
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "/Users/gallaghe/Code/nutrition-labels\n"
+     ]
+    }
+   ],
+   "source": [
+    "cd .."
```
Comment (aCampello) [inline notebooks/Ensemble parameter exploration.ipynb:68]: ```cd ..`` ` might not be ideal (for example if you happen to run the notebook twice, it won't work, making the code not idempotent). Probably better to use the full paths when reading the file.

File: notebooks/Ensemble parameter exploration.ipynb [hunk 7]
```diff
@@ -0,0 +1,1124 @@
...
+       "9                            0  218358/Z/19/Z  "
+      ]
+     },
+     "execution_count": 5,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# The test predictions for each model\n",
```
Comment (aCampello) [inline notebooks/Ensemble parameter exploration.ipynb:544]: I had the impression that  you should en up with `2**12` models (each model included/not included), but the data frame only has 245. Any thoughts?

Comment (aCampello) [reply notebooks/Ensemble parameter exploration.ipynb:544]: Oh, I see, you test all the combinations below. My bad, ignore it!

File: docs/Prodigy_training_data.md [hunk 8]
```diff
@@ -79,63 +79,47 @@ python nutrition_labels/prodigy_training_data.py --prodigy_data_dir 'data/prodig
...
-| 1 | Relevant | 592 |
-| 0 | Not relevant | 576 |
+| 1 | Relevant | 495 |
+| 0 | Not relevant | 485 |
 
 ### Variability in the results:
 
-I reran `grant_tagger_seed_experiments.py` with the training data in `data/processed/training_data/210218/training_data.csv`. This script trains several models with different random seeds and teamed with the `Seed variability.ipynb` notebook we can pick a new 'best' random seed to choose our training/test split for all models.
+
+I reran `grant_tagger_seed_experiments.py` with the training data in `data/processed/training_data/210221/training_data.csv`. This script trains several models with different random seeds and teamed with the `Seed variability.ipynb` notebook we can pick a new 'best' random seed to choose our training/test split for all models.
```
Comment (nsorros) [inline docs/Prodigy_training_data.md:92]: It is worth pausing and thinking whether picking the best seed is a good idea. I think researchers do that when they publish results but ideally what you want is low variance to begin with and any model not high variance and the best model there as this overfits the validation/test data.

I would like to hear @aCampello thoughts here.

File: docs/Prodigy_training_data.md [hunk 9]
```diff
@@ -145,53 +129,83 @@ A comparison with the previous results (201022 - old definition, 210128 - new de
...
+The ensemble parameter experiments were run in `notebooks/Ensemble parameter exploration.ipynb`. In this we plot the precision/recall for each of the 12 models plus every single combination of those 12 models for different values of the number of models that need to agree.
+
+![](figures/model_performance_ensemble_vs_single.png)
+
+The models found with the 'best' recall and precision were:
+
+- Ensemble of 3 models with 2 needing to agree ('bert_log_reg_bert_210221', 'bert_SVM_bert_210221', 'count_naive_bayes_210221')
+- Ensemble of 4 models with 3 needing to agree ('bert_naive_bayes_bert_210221', 'bert_log_reg_bert_210221', 'bert_SVM_bert_210221', 'count_naive_bayes_210221')
+- Ensemble of 4 models with 3 needing to agree ('bert_log_reg_bert_210221', 'bert_SVM_bert_210221', 'bert_naive_bayes_scibert_210221', 'count_naive_bayes_210221')
+- Ensemble of 3 models with 4 needing to agree ('bert_naive_bayes_bert_210221', 'bert_log_reg_bert_210221', 'bert_SVM_bert_210221', 'bert_naive_bayes_scibert_210221', 'count_naive_bayes_210221')
```
Comment (nsorros) [inline docs/Prodigy_training_data.md:188]: 3 modes with 4 needed to agree?

File: docs/Prodigy_training_data.md [hunk 10]
```diff
@@ -145,53 +129,83 @@ A comparison with the previous results (201022 - old definition, 210128 - new de
...
+
+The ensemble parameter experiments were run in `notebooks/Ensemble parameter exploration.ipynb`. In this we plot the precision/recall for each of the 12 models plus every single combination of those 12 models for different values of the number of models that need to agree.
+
+![](figures/model_performance_ensemble_vs_single.png)
+
+The models found with the 'best' recall and precision were:
+
+- Ensemble of 3 models with 2 needing to agree ('bert_log_reg_bert_210221', 'bert_SVM_bert_210221', 'count_naive_bayes_210221')
+- Ensemble of 4 models with 3 needing to agree ('bert_naive_bayes_bert_210221', 'bert_log_reg_bert_210221', 'bert_SVM_bert_210221', 'count_naive_bayes_210221')
+- Ensemble of 4 models with 3 needing to agree ('bert_log_reg_bert_210221', 'bert_SVM_bert_210221', 'bert_naive_bayes_scibert_210221', 'count_naive_bayes_210221')
```
Comment (nsorros) [inline docs/Prodigy_training_data.md:187]: it is worth saying that this ensemble technique is called majority voting and you can use VotingClassifier to easily implement it https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.VotingClassifier.html

Comment (nsorros) [reply docs/Prodigy_training_data.md:187]: additionally if you want to extend the voting classifier to specify a more strict majority voting i.e. instead of simple majority a preselected number of models, I would recommend adding the extension to WellcomeML. I have already extended VotingClassifier to receive pretrained models https://github.com/wellcometrust/WellcomeML/blob/master/wellcomeml/ml/voting_classifier.py, you can introduce further parameters to do what you want

File: notebooks/Ensemble parameter exploration.ipynb [hunk 11]
```diff
@@ -0,0 +1,1124 @@
...
+     "execution_count": 9,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# Get every combination of models\n",
+    "models_list = model_results.index.tolist()\n",
+    "all_combinations = []\n",
+    "for r in range(2, len(models_list) + 1):\n",
```
Comment (aCampello) [inline notebooks/Ensemble parameter exploration.ipynb:674]: why 2? I think it should be 1, so you end up with `2**12-1` models.

File: docs/Prodigy_training_data.md [hunk 12]
```diff
@@ -145,53 +129,83 @@ A comparison with the previous results (201022 - old definition, 210128 - new de
...
+
+||precision|recall|f1-score|support|
+|--|--|--|--|--|
+|0|0.78|0.82|0.80|110|
+|1|0.85|0.81|0.83|135|
+|accuracy|||0.82|245|
+|macro avg|0.81|0.82|0.82|245|
+|weighted avg|0.82|0.82|0.82|245|
+
+Finally these details were put into a config file and
```
Comment (nsorros) [inline docs/Prodigy_training_data.md:206]: ok this is where the config files comes from. note that it is better for reproducibility to submit configs on all models you report results in, not just the ones you "productionise"

File: nutrition_labels/tech_grant_tagger.py [hunk 13]
```diff
@@ -73,31 +50,49 @@ def clean_grants_data(self, previous_grant_data):
...
         grant_data = self.clean_grants_data(grant_data)
         grants_text = grant_data['Grant texts'].tolist()
         self.grants_ids = grant_data[self.grant_id_col].tolist()
 
         return grants_text
 
+    def load_model(self, model_path):
+
+        # Loading a trained model and vectorizer to predict on all the grants data:
+        grant_tagger_loaded = GrantTagger()
```
Comment (nsorros) [inline nutrition_labels/tech_grant_tagger.py:67]: careful here, my project is also called grants_tagger :p

Comment (nsorros) [reply nutrition_labels/tech_grant_tagger.py:67]: how is `GrantTagger` different from `TechGrantModel`, why there are two classes?

File: notebooks/Ensemble parameter exploration.ipynb [hunk 14]
```diff
@@ -0,0 +1,1124 @@
...
+      ]
+     },
+     "execution_count": 22,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# Less models to have to make predictions from the better\n",
+    "optimal_model_i = 411\n",
```
Comment (aCampello) [inline notebooks/Ensemble parameter exploration.ipynb:1032]: I don't get it where 411 comes from?

File: docs/Prodigy_training_data.md (hunk 8 above)
Comment (aCampello) [reply docs/Prodigy_training_data.md:92]: I haven't followed this very closely, so I am not sure what the cross validation technique is, but I suspect that if a `n-fold` (say 5-fold) validation is used, the effect of the seed will be smoothed. I agree that people use it to guarantee that your results are replicable, not intended as a hyper parameter.

From what I saw from the notebook, the effect of the random seed is like 1%-ish, which is probably ok?

File: requirements.txt (hunk 1 above)
Comment (aCampello) [reply requirements.txt:11]: Nice!

**Review by aCampello — APPROVED** [review]
I can't say I got a full understanding of all steps, but the notebook runs for me and outputs the same results, with some minor observations. 

Worth thinking about Nick's observation on random seeds.

File: Makefile (hunk 5 above)
Comment (lizgzil) [reply Makefile:76]: basically - just copying to and from S3

File: Makefile (hunk 4 above)
Comment (lizgzil) [reply Makefile:67]: I was working out a way to allow us to not have to sync the entire models/data folder for PRs (which are big!) so `sync_latest_files_to_s3` will copy the files/folders listed in `latest_files.txt`. It's line by line since I don't want to sync entire folders.
Did you try `make sync_latest_files_from_s3`, would be interested to know if it's worked correctly?

File: configs/ensemble/2021.02.20.ini (hunk 3 above)
Comment (lizgzil) [reply configs/ensemble/2021.02.20.ini:1]: it is for the ensemble model as this PR is the one that introduces them. But we have them for creating the training data and cluster analysis too

File: docs/Prodigy_training_data.md (hunk 8 above)
Comment (lizgzil) [reply docs/Prodigy_training_data.md:92]: yes a good point about overfitting to the test data, especially since the dataset is so small too.

File: docs/Prodigy_training_data.md (hunk 10 above)
Comment (lizgzil) [reply docs/Prodigy_training_data.md:187]: ah yes - I was under the impression that our vectorizers (not all being from sklearn) would cause problems with VotingClassifier, but I will revisit that.

File: docs/Prodigy_training_data.md (hunk 9 above)
Comment (lizgzil) [reply docs/Prodigy_training_data.md:188]: oops! should be 5

File: Makefile (hunk 4 above)
Comment (nsorros) [reply Makefile:67]: did not try but good idea. I agree we need to improve on selectively syncing files. maybe dvc is the way but what are doing also makes sense. 👍

File: docs/Prodigy_training_data.md (hunk 10 above)
Comment (nsorros) [reply docs/Prodigy_training_data.md:187]: I think our vectorisers should work as we have implemented them to be sklearn compatible, if they are not we should fix that.

**Review by nsorros — APPROVED** [review]
LGTM, minor comments. Would love to see VotingClassifier being used from WellcomeML with improvements from this project

File: nutrition_labels/prodigy_training_data.py (hunk 2 above)
Comment (lizgzil) [reply nutrition_labels/prodigy_training_data.py:46]: yep :( the results of 210218 which are all commented out are from this bug

File: nutrition_labels/tech_grant_tagger.py (hunk 13 above)
Comment (lizgzil) [reply nutrition_labels/tech_grant_tagger.py:67]: yes, good point. GrantTagger is training single models (e.g. Count+SVM). TechGrantModel is combining the trained models to make ensemble predictions.

Comment (lizgzil) [reply nutrition_labels/tech_grant_tagger.py:67]: maybe TechGrantModel-> EnsembleGrantTagger?

File: notebooks/Ensemble parameter exploration.ipynb (hunk 11 above)
Comment (lizgzil) [reply notebooks/Ensemble parameter exploration.ipynb:674]: If r=1 then the "combination" would just be the original model by itself - which I have the test results for anyway.

File: notebooks/Ensemble parameter exploration.ipynb (hunk 14 above)
Comment (lizgzil) [reply notebooks/Ensemble parameter exploration.ipynb:1032]: oh yes, I'll add my logic here instead of hardcoding. I've just manually picked which my favourite combination from the dataframe (row 411). But that was because it has the fewest number of models, so I can add this logic in instead

File: nutrition_labels/tech_grant_tagger.py (hunk 13 above)
Comment (nsorros) [reply nutrition_labels/tech_grant_tagger.py:67]: I see, yeah that is a better name imo

**Comment by lizgzil** [comment]
I've opened https://github.com/wellcometrust/nutrition-labels/issues/58 and https://github.com/wellcometrust/nutrition-labels/issues/59
to revisit the random seed and the voting classifier methodology.
